### ISSUE TITLE
Should AudioLDM pipeline use separate unet class

### ISSUE THREAD

**Issue by williamberman** [huggingface/diffusers#4658]
@patrickvonplaten see comment here https://github.com/huggingface/diffusers/pull/4549/files#r1297472583 and the rationale in the corresponding docstring. Would be great to get you to sanity check. Thanks!

**Comment by sanchit-gandhi** [comment]
Copying the Slack discussion we had offline for context:

**Sanchit**
Into the fiddly bit of the UNet - would love the second opinion of the diffusers maintainers here!

The vanilla UNet cross-attention layer looks as follows:
* Resnet Block (hidden_states)
* Cross-Attention Transformer Block (hidden_states, encoder_hidden_states)

=> this is the architecture that is used in diffusers when we run the UNet forward with arguments of (hidden_states, encoder_hidden_states)

AudioLDM 2 extends the vanilla UNet architecture to use an additional self-attention layer and two cross-attention layers:
* Resnet Block (hidden_states)
* Self-Attention Transformer Block with double self-attention (hidden_states)
* Cross-Attention Transformer Block 1 (hidden_states, encoder_hidden_states_1)
* Cross-Attention Transformer Block 2 (hidden_states, encoder_hidden_states_2)

=> here we use a different set of encoder hidden-states for cross-attention blocks 1 and 2 (they are obtained from different conditioning models). Also, we don’t want to pass either of these encoder hidden-states to the self-attention layer, since it uses double self-attention

The naive way of integrating these changes involves adding an extra argument encoder_hidden_states_2 to the UNet, and propagating this new argument forward to each of the transformer layers. Is there a clean way we can build on the current UNet in diffusers without adding a bunch of extra arguments and control flow throughout the code? E.g. with callbacks?

**Sanchit**

Note also that the two encoder hidden-states require different attention masks, so this adds another argument for encoder_attention_mask_2


**Sayak Paul**

I am not too familiar with the callback mechanism. Maybe create a new model class subclassing from the original UNet model?


**Sanchit**

Can do! Will try to do this in a way where there’s minimal new functionality, such that AudioLDM2 receives any updates to the base UNet

**Sanchit**

Ooof because of the dependency structure in diffusers it looks like we’ll have to override the `__init__` and `forward` for quite a large number of models to accommodate the new model structure and extra arguments:
* UNet2DConditionModel
* CrossAttnDownBlock2D + UNetMidBlock2DCrossAttn + CrossAttnUpBlock2D
* Transformer2DModel

=> it’s going to be quite a big effort to add all of these standalone for AudioLDM2 (edited) 

**Sanchit**

I had thought maybe we could pass the extra cross-attention arguments using the cross_attention_kwargs, but I don’t think this is a good idea passing a dict of tensors to the model (will make it hard to torch.compile)

**Sayak Paul**

kwargs should be restricted to scalar value arguments IMO and as you correctly pointed out otherwise, it might introduce graph breaks in torch.compile().

**Sanchit**

Trying quickly to put together a quick PoC to see what it would look like if we integrated these changes directly into the base UNet. It’s faster this way than adding all the code standalone. If we don’t like it in the base UNet, I’ll pivot to adding it standalone

**Will Berman**

I don't follow the entirety but adding new models is also an option. I think we try to avoid extending existing models if possible


**Will Berman**

We have a few existing one off unet blocks that are only used by only one model


**Sanchit**

Okay cool - made all the changes in the existing models so far, but should be easy to copy these into a standalone UNet 


**Sanchit**

Thanks both for your help - refactored as suggested above to use a standalone model!

**Comment by patrickvonplaten** [comment]
Thanks for putting up the issue :-)

Standalone model is ok for me here! Think we can revisit this decision if AudioLDM is getting lots of traction. Just some comments:

> Sanchit

> I had thought maybe we could pass the extra cross-attention arguments using the cross_attention_kwargs, but I don’t think this is a good idea passing a dict of tensors to the model (will make it hard to torch.compile)

> Sayak Paul

> kwargs should be restricted to scalar value arguments IMO and as you correctly pointed out otherwise, it might introduce graph breaks in torch.compile().

I don't see how passing a dict of tensors is not a problem for torch.compile IMO - we're already doing this for SDXL here: https://github.com/huggingface/diffusers/blob/f22aad6e3add72a5f2bab067a117dbeaea33d540/src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py#L811 - why should this break the graph?
